### ISSUE TITLE
Refactor `thrust::reduce` to use `cub::DeviceReduce`

### ISSUE THREAD

**Issue by elstehle** [NVIDIA/cccl#1626]

**Comment by bernhardmgruber** [comment]
The port of Thrust to use CUB is attempted here: https://github.com/NVIDIA/cccl/pull/4970. I am reusing `thrust::reduce_into`.

**Comment by bernhardmgruber** [comment]
I don't think we need a fallback policy for DeviceReduce. We run out of SMEM backed temporary storage really late. Here is a table for sm86:
```
// sizeof(value_t) -> sizeof(AgentReduce<...>::TempStorage)
// 16 -> 224
// 32 -> 368
// 64 -> 656
// 4Ki -> 8216
// 16Ki -> 32792
// 24Ki -> 49176
// 32Ki -> 65560 (but causes Error: Formal parameter space overflowed)
```
I think reducing values where each value is close to 24KiB is extremely rare.

**Comment by bernhardmgruber** [comment]
I think to properly port to `cub::DeviceReduce` we want to:

* Extend `cub::DeviceReduce::ArgMin|Max` with a custom less then operator
* Extend `cub::DeviceReduce::ArgMin|Max` to use a `FutureValue` to the first element in case `::cuda::std::numeric_limits` is not specialized for the input type, or a comparison predicate other than `cuda::std::less` is used.
* Provide a combined `cub::DeviceReduce::ArgMinMax`

**Comment by bernhardmgruber** [comment]
>  Add sanity tests for large number of items for `thrust/extrema.h`

I think this was added in 9b164e2d.
